pkg/plugins/manager/loader/loader.go

```diff
@@ -195,6 +203,16 @@ func (l *Loader) loadPlugins(ctx context.Context, class plugins.Class, pluginJSO
 			return nil, err
 		}
 		metrics.SetPluginBuildInformation(p.ID, string(p.Type), p.Info.Version, string(p.Signature))
+
+		if l.features.IsEnabled(featuremgmt.FlagAccessControlOnCall) && len(p.JSONData.Roles) > 0 {
+			if errDeclareRoles := l.acService.DeclarePluginRoles(p.ID, p.JSONData.Roles...); errDeclareRoles != nil {
+				l.log.Warn("Declare plugin roles failed",
+					"pluginID", p.ID,
+					"warning", "Make sure the role declaration is correct.",
+					"path", p.PluginDir+"/plugin.json",
+					"error", errDeclareRoles)
+			}
+		}
```

Comment:
We're trying to make /pkg/plugins (plugin management) and most of it's sub-packages to not have any dependencies to the rest of Grafana backend. Why you might ask? To 
make plugin management loosely coupled and to allow a clear separation between Grafana and plugin management which could allow running plugin management as a separate service in a possible future.  

Possibly somewhat related to some of the ideas with  https://en.wikipedia.org/wiki/Hexagonal_architecture_(software)

What are the alternatives:
- Plugin management to define it's own internal interface for registering roles and then it would be up to Grafana (access control) to provide an implementation of that interface.
  - Benefits
    - roles would be registered when any plugin are installed/loaded
  - Drawbacks
    - some additional abstractions/work within plugin management/access control
- Grafana to register plugin roles at startup and when installing plugins during runtime
  - Could call access control somewhere around here: https://github.com/grafana/grafana/blob/05ceff5188d97e617668e84c820d0c7af8a2b934/pkg/server/server.go#L132-L134
  - And could call access control in this handler, after a plugin have been installed: https://github.com/grafana/grafana/blob/05ceff5188d97e617668e84c820d0c7af8a2b934/pkg/api/plugins.go#L409
  - Drawbacks
    - If a plugin is installed from within another service it would not register the plugin roles out of the box
- Access control to listen for plugins that have been installed and read plugin metadata and register roles based on that.
- Using the in-proc publish of event functionality: https://github.com/grafana/grafana/blob/05ceff5188d97e617668e84c820d0c7af8a2b934/pkg/bus/bus.go#L23-L26
  - Benefits
    - Good separation of concerns. Plugin management just have to care about the plugin.json/meta data contract, but not exactly what it's used for
  - Drawbacks
    - Only works in an in-proc world

Reply:
So as we discussed, we are going for the first approach you suggested, having an interface defined in "plugin management" that access-control implements.

I just have a small problem with removing the `io.ReadCloser` on the `plugin.json` in favor of the already parsed DTO, in `/pkg/services/plugins/accesscontrol.go`, we have a dependency to the `accesscontrol` package, that prevents me from using types defined in `/pkg/services/plugins/models.go`.
I can move the file `/pkg/services/plugins/accesscontrol.go` somewhere else, maybe into its own package  `/pkg/services/plugins/pluginac/accesscontrol.go`, or move the newly defined structs for `RoleRegistration` in such a package, but it's not ideal.
The `io.ReadCloser` offers the benefit of truly decoupling both packages for the future changes you mentioned, but I agree that it has the drawback of reading and parsing the same file twice.

What do you think?

Reply:
Wait, I might be onto something in the end.

Reply:
Okay :smile: 

In general, I find roles/actions being defined in multiple places today, but still many being defined in https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/models.go. Would it make sense to have the plugin related accesscontrol types/constants there as well?

Reply:
That's a good question, we have gone back and forth on this subject. :confused: 
I personally now like to see permissions related to a kind of resource defined in the package managing that resource. But we have a clear example here of why this might not be ideal.
I'm open to discussion but I'd rather not modify this in this PR :slightly_smiling_face: